core/src/impl/Kokkos_Atomic_Exchange.hpp

```diff
@@ -419,6 +419,124 @@ void atomic_assign( volatile T * const dest_v , const T val )
 #endif
 #endif
 
+
+#if defined( KOKKOS_ENABLE_HIP )
+__inline__ __device__
+int atomic_exchange( volatile int * const dest , const int val )
+{
+  return atomicExch( static_cast<int*>(dest) , val );
+}
+
+__inline__ __device__
+unsigned int atomic_exchange( volatile unsigned int * const dest , const unsigned int val )
+{
+  return atomicExch( static_cast<unsigned int*>(dest) , val );
+}
+
+__inline__ __device__
+unsigned long long int atomic_exchange( volatile unsigned long long int * const dest , const unsigned long long int val )
+{
+  // return __ullAtomicExch( (unsigned long long*) dest , val );
+  return atomicExch( (unsigned long long*) dest , val );
+}
+
+/** \brief  Atomic exchange for any type with compatible size */
+template< typename T >
+__inline__ __device__
+T atomic_exchange(
+  volatile T * const dest ,
+  typename Kokkos::Impl::enable_if< sizeof(T) == sizeof(int) , const T & >::type val )
+{
+
+  int tmp = atomicExch( ((int*)dest) , *((int*)&val) );
+  return *((T*)&tmp);
```

Comment:
`static_cast<T&>( ... )` and `const_cast` instead of C-style conversion for the args